server/src/main/java/org/opensearch/search/aggregations/bucket/FastFilterRewriteHelper.java

```diff
@@ -439,61 +402,297 @@ public static boolean tryFastFilterAggregation(
+                        collector.count();
+                    }
+                }
+            }
+
+            @Override
+            public PointValues.Relation compare(byte[] minPackedValue, byte[] maxPackedValue) {
+                long min = NumericUtils.sortableBytesToLong(minPackedValue, 0);
+                long max = NumericUtils.sortableBytesToLong(maxPackedValue, 0);
+                long queryMin = collector.activeRange[0];
+                long queryMax = collector.activeRange[1];
+
+                boolean crosses = false;
+                if (collector.activeRange[1] < min) {
+                    // need to move to next range
+                    collector.finalizePreviousRange();
+                    if (collector.iterateRangeEnd(min)) {
+                        throw new CollectionTerminatedException();
+                        // return PointValues.Relation.CELL_OUTSIDE_QUERY;
+                    }
+
+                    // compare the next range with this node's min max again
+                    // it cannot be outside again, can only be cross or inside
+                    if (collector.activeRange[1] < max) {
+                        crosses = true;
+                    }
+                } else if (queryMin > min || queryMax < max) {
+                    crosses = true;
+                }
+
+                if (crosses) {
+                    return PointValues.Relation.CELL_CROSSES_QUERY;
+                } else {
+                    return PointValues.Relation.CELL_INSIDE_QUERY;
+                }
+            }
+        };
+    }
+
+    private static class RangeCollectorForPointTree {
+        private final BiConsumer<Long, Integer> incrementDocCount;
+        private final DateFieldMapper.DateFieldType fieldType;
+        private int counter = 0;
+        private long[] activeRange;
+        private final Iterator<long[]> rangeIter;
+        private int visitedRange = 0;
+        private final int size; // the given size of non-zero buckets used in composite agg
+
+        public RangeCollectorForPointTree(
+            BiConsumer<Long, Integer> incrementDocCount,
+            DateFieldMapper.DateFieldType fieldType,
+            Iterator<long[]> rangeIter,
+            int size
+        ) {
+            this.incrementDocCount = incrementDocCount;
+            this.fieldType = fieldType;
+            this.rangeIter = rangeIter;
+            this.size = size;
         }
 
-        logger.debug("Fast filter optimization applied to shard {} segment {}", fastFilterContext.context.indexShard().shardId(), ctx.ord);
-        return true;
+        private void count() {
+            counter++;
+        }
+
+        private void finalizePreviousRange() {
+            if (counter > 0) {
+                incrementDocCount.accept(fieldType.convertNanosToMillis(activeRange[0]), counter);
+                counter = 0;
+            }
+        }
```

Comment:
It seems we are always incrementing the counter by 1 unit. This needs to be incremented by the number of documents in each of the bucket which kind of becomes O(n). How is this faster than the original implementation? Am I missing something?

Reply:
Right, for the node that's inside the range, we can just get the size of this node, no need to visit every point of the node. 
Later on for supporting the sub aggregation, I plan to collect the docID of each point, so sub aggregation can collect on these docIDs.

I will do benchmark around how much difference between visit every point vs. get the size. I would expect a noticeable difference.

Reply:
> Right, for the node that's inside the range, we can just get the size of this node, no need to visit every point of the node.

Currently that logic is missing, right!? So, the code changes are still in progress?

> Later on for supporting the sub aggregation, I plan to collect the docID of each point, so sub aggregation can collect on these docIDs.

Will be great if the same approach can be leveraged for sub aggregation as well!

> I will do benchmark around how much difference between visit every point vs. get the size. I would expect a noticeable difference.

I will expect significant improvement as well!

Reply:
Added the logic to directly get size from node.